x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/service/TransportCreateServiceAccountTokenAction.java

```diff
@@ -18,37 +18,31 @@
 import org.elasticsearch.xpack.core.security.action.service.CreateServiceAccountTokenRequest;
 import org.elasticsearch.xpack.core.security.action.service.CreateServiceAccountTokenResponse;
 import org.elasticsearch.xpack.core.security.authc.Authentication;
-import org.elasticsearch.xpack.security.authc.service.IndexServiceAccountTokenStore;
-import org.elasticsearch.xpack.security.authc.support.HttpTlsRuntimeCheck;
+import org.elasticsearch.xpack.security.authc.service.ServiceAccountService;
 
 public class TransportCreateServiceAccountTokenAction
     extends HandledTransportAction<CreateServiceAccountTokenRequest, CreateServiceAccountTokenResponse> {
 
-    private final IndexServiceAccountTokenStore indexServiceAccountTokenStore;
+    private final ServiceAccountService serviceAccountService;
     private final SecurityContext securityContext;
-    private final HttpTlsRuntimeCheck httpTlsRuntimeCheck;
 
     @Inject
     public TransportCreateServiceAccountTokenAction(TransportService transportService, ActionFilters actionFilters,
-                                                    IndexServiceAccountTokenStore indexServiceAccountTokenStore,
-                                                    SecurityContext securityContext,
-                                                    HttpTlsRuntimeCheck httpTlsRuntimeCheck) {
+                                                    ServiceAccountService serviceAccountService,
+                                                    SecurityContext securityContext) {
         super(CreateServiceAccountTokenAction.NAME, transportService, actionFilters, CreateServiceAccountTokenRequest::new);
-        this.indexServiceAccountTokenStore = indexServiceAccountTokenStore;
+        this.serviceAccountService = serviceAccountService;
         this.securityContext = securityContext;
-        this.httpTlsRuntimeCheck = httpTlsRuntimeCheck;
     }
 
     @Override
     protected void doExecute(Task task, CreateServiceAccountTokenRequest request,
                              ActionListener<CreateServiceAccountTokenResponse> listener) {
-        httpTlsRuntimeCheck.checkTlsThenExecute(listener::onFailure, "create service account token", () -> {
-            final Authentication authentication = securityContext.getAuthentication();
-            if (authentication == null) {
-                listener.onFailure(new IllegalStateException("authentication is required"));
-            } else {
-                indexServiceAccountTokenStore.createToken(authentication, request, listener);
-            }
-        });
+        final Authentication authentication = securityContext.getAuthentication();
+        if (authentication == null) {
+            listener.onFailure(new IllegalStateException("authentication is required"));
+        } else {
+            serviceAccountService.createIndexToken(authentication, request, listener);
+        }
```

Comment:
I don't follow why this change is here (to move things from the actions to the service).

I don't have any opinion on it - I'm just confused about where it came from. 
I assume it's a consequence of having the Nodes action that needs to query for index & file tokens separately, but I can't quite piece all the bits together.

Reply:
You understanding is correct. Please let me fill in some more details. 

The logic is not really moved from the "action" to the service, but rather moved from `IndexServiceAccountTokenStore` to the service. Because of the separation between index-token and file-token fetching, the service becomes more aware of the difference. Previously I tried to make the service not aware of the underlying difference between different stores by having just a single `CompositeServiceAccountTokenStore`. But this is no long suitable because we have to handle token fetching separately. 

Now the service knows explicitly about both stores and can hence become a facade for both of them. So I think it is better have the abstraction layer here since it is already aware of it. That's why I decided to let the actions call into service instead of using the underlying store directly. I think in general I prefer this layer separation, but we are (I am) not consistent about it.

Since this action now uses the service and the service already handles HTTP TLS check, it is also moved into the service. One benefit is that I was able to test it in one place and delete a few tests for the actions.